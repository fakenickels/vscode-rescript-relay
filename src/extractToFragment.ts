import {
  DocumentNode,
  getLocation,
  getNamedType,
  GraphQLSchema,
  SelectionNode,
  SelectionSetNode,
  Source,
  TypeInfo,
  visit,
  visitWithTypeInfo,
} from "graphql";
import { Position } from "vscode";

interface ExtractToFragmentConfig {
  normalizedSelection: [Position, Position];
  parsedOp: DocumentNode;
  schema: GraphQLSchema;
  source: Source;
}
export const extractToFragment = ({
  normalizedSelection,
  parsedOp,
  schema,
  source,
}: ExtractToFragmentConfig): null | {
  selections: SelectionNode[];
  targetSelection: SelectionSetNode;
  parentTypeName: string;
} => {
  // Make selection into fragment component
  let parentTypeName: string | undefined;
  let targetSelection: SelectionSetNode | undefined;

  const [st, en] = normalizedSelection;

  const typeInfo = new TypeInfo(schema);

  const visitor = visitWithTypeInfo(typeInfo, {
    SelectionSet(node, _key, _parent, _t, _c) {
      const { loc } = node;

      if (!loc) {
        return;
      }

      const start = getLocation(source, loc.start);
      const end = getLocation(source, loc.end);

      if (st.line >= start.line && en.line <= end.line) {
        const thisType = typeInfo.getType();

        if (thisType) {
          parentTypeName = getNamedType(thisType).name;
          targetSelection = node;
        }
      }
    },
  });

  visit(parsedOp, visitor);

  const selections: SelectionNode[] = targetSelection
    ? targetSelection.selections.filter((s: SelectionNode) => {
        if (s.loc) {
          const sLocStart = getLocation(source, s.loc.start);
          const sLocEnd = getLocation(source, s.loc.end);

          return sLocStart.line >= st.line && sLocEnd.line <= en.line;
        }

        return false;
      })
    : [];

  if (!targetSelection || !parentTypeName) {
    return null;
  }

  return { selections, targetSelection, parentTypeName };
};
