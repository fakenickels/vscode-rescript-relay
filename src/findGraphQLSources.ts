import { getLocator } from "locate-character";
import { GraphQLSource, GraphQLSourceFromTag } from "./extensionTypes";

/**
 * A helper for extracting GraphQL operations from source via a regexp.
 * It assumes that the only thing the regexp matches is the actual content,
 * so if that's not true for your regexp you probably shouldn't use this
 * directly.
 */
export let makeExtractTagsFromSource = (
  regexp: RegExp
): ((text: string) => Array<GraphQLSourceFromTag>) => (
  text: string
): Array<GraphQLSourceFromTag> => {
  const locator = getLocator(text);
  const sources: Array<GraphQLSourceFromTag> = [];
  let result;
  while ((result = regexp.exec(text)) !== null) {
    let start = locator(result.index);
    let end = locator(result.index + result[0].length);

    sources.push({
      type: "TAG",
      content: result[0],
      start: {
        line: start.line,
        character: start.column,
      },
      end: {
        line: end.line,
        character: end.column,
      },
    });
  }

  return sources;
};

export const jsGraphQLTagsRegexp = new RegExp(
  /(?<=(graphql|gql|graphql\.experimental)`)[.\s\S]+?(?=`)/g
);
export const rescriptFileFilterRegexp = new RegExp(/(\%relay\()/g);
export const rescriptGraphQLTagsRegexp = new RegExp(
  /(?<=\%relay\([\s]*`)[\s\S.]+?(?=`[\s]*\))/g
);

const extractGraphQLSourceFromJs = makeExtractTagsFromSource(
  jsGraphQLTagsRegexp
);

export const extractGraphQLSourceFromReScript = makeExtractTagsFromSource(
  rescriptGraphQLTagsRegexp
);

export function extractGraphQLSources(
  languageId: string,
  document: string
): GraphQLSource[] | null {
  switch (languageId) {
    case "graphql":
      return [
        {
          type: "FULL_DOCUMENT",
          content: document,
        },
      ];
    case "javascript":
    case "javascriptreact":
    case "typescript":
    case "typescriptreact":
    case "vue":
      return extractGraphQLSourceFromJs(document);
    case "rescript":
      return extractGraphQLSourceFromReScript(document);
    default:
      return null;
  }
}

export function extractSelectedOperation(
  languageId: string,
  document: string,
  selection: {
    line: number;
    character: number;
  }
): GraphQLSource | null {
  const sources = extractGraphQLSources(languageId, document);

  if (!sources || sources.length < 1) {
    return null;
  }

  let targetSource: GraphQLSource | null = null;

  if (sources[0].type === "FULL_DOCUMENT") {
    targetSource = sources[0];
  } else {
    // A tag must be focused
    for (let i = 0; i <= sources.length - 1; i += 1) {
      const t = sources[i];

      if (
        t.type === "TAG" &&
        selection.line >= t.start.line &&
        selection.line <= t.end.line
      ) {
        targetSource = t;
      }
    }
  }

  return targetSource;
}

export function getSelectedGraphQLOperation(
  doc: string,
  pos: { line: number; character: number }
): GraphQLSourceFromTag | null {
  const selectedOperation = extractSelectedOperation("rescript", doc, {
    line: pos.line,
    character: pos.character,
  });

  if (selectedOperation && selectedOperation.type === "TAG") {
    return selectedOperation;
  }

  return null;
}
