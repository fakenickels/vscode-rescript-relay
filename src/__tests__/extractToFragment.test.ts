import { parse, Source, print, buildSchema, SelectionNode } from "graphql";
import { extractToFragment } from "../extractToFragment";
import { getSelectedGraphQLOperation } from "../findGraphQLSources";

const makeMockNormalizedSelection = (
  startLine: number,
  endLine: number
): [any, any] => [{ line: startLine }, { line: endLine }];

const printExtractedFragment = (
  parentTypeName: string,
  selections: SelectionNode[]
) =>
  print({
    kind: "FragmentDefinition",
    typeCondition: {
      kind: "NamedType",
      name: { kind: "Name", value: parentTypeName },
    },
    name: { kind: "Name", value: "SomeFragment_user" },
    selectionSet: {
      kind: "SelectionSet",
      selections: selections,
    },
  });

const mockSchema = `
type User {
    id: ID!
    firstName: String!
    lastName: String!
    bestFriend: Friend!
}

type Friend {
    id: ID!
    avatarUrl: String!
    friendsSince: Int!
    status: FriendStatus
}

type FriendStatus {
  since: Int!
  from: String
  to: String
}
`;

const schema = buildSchema(mockSchema);

const targetSource = `// Other stuff

module Fragment = %relay(\`
  fragment SomeFragment_user on User {
    id
    firstName
    lastName
    bestFriend {
        id
        avatarUrl
        friendsSince
        status {
            since
            from 
            to
        }
    }
  }
\`
)

// Some other source here
      `;

const selectedOp = getSelectedGraphQLOperation(targetSource, {
  line: 6,
  character: 2,
} as any);

if (!selectedOp) {
  throw new Error("Select an op please..");
}

const parsedOp = parse(selectedOp.content);
const source = new Source(selectedOp.content);

describe("Extract to fragment component", () => {
  it("extracts in simple cases", () => {
    const extractedFragment = extractToFragment({
      schema,
      normalizedSelection: makeMockNormalizedSelection(4, 6),
      parsedOp,
      source,
    });

    if (!extractedFragment) {
      throw new Error("Could not extract fragment.");
    }

    expect(
      printExtractedFragment(
        extractedFragment.parentTypeName,
        extractedFragment.selections
      )
    ).toMatchSnapshot();
  });

  it("extracts in slightly more advanced examples", () => {
    const extractedFragment = extractToFragment({
      schema,
      normalizedSelection: makeMockNormalizedSelection(8, 15),
      parsedOp,
      source,
    });

    if (!extractedFragment) {
      throw new Error("Could not extract fragment.");
    }

    expect(
      printExtractedFragment(
        extractedFragment.parentTypeName,
        extractedFragment.selections
      )
    ).toMatchSnapshot();
  });

  it("extracts even when nested quite far", () => {
    const extractedFragment = extractToFragment({
      schema,
      normalizedSelection: makeMockNormalizedSelection(12, 13),
      parsedOp,
      source,
    });

    if (!extractedFragment) {
      throw new Error("Could not extract fragment.");
    }

    expect(
      printExtractedFragment(
        extractedFragment.parentTypeName,
        extractedFragment.selections
      )
    ).toMatchSnapshot();
  });
});
