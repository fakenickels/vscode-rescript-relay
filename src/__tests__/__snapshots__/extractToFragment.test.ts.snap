// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extract to fragment component extracts even when nested quite far 1`] = `
"fragment SomeFragment_user on FriendStatus {
  from
  to
}"
`;

exports[`Extract to fragment component extracts in simple cases 1`] = `
"fragment SomeFragment_user on User {
  firstName
  lastName
}"
`;

exports[`Extract to fragment component extracts in slightly more advanced examples 1`] = `
"fragment SomeFragment_user on Friend {
  avatarUrl
  friendsSince
  status {
    since
    from
    to
  }
}"
`;
